File: app/src/main/java/com/cardee/owner_home/view/OwnerCarsFragment.java
package com.cardee.owner_home.view;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.cardee.R;
import com.cardee.domain.owner.entity.Car;
import com.cardee.owner_home.OwnerCarListContract;
import com.cardee.owner_home.presenter.OwnerCarsPresenter;
import com.cardee.owner_home.view.adapter.CarListAdapter;
import com.cardee.owner_home.view.listener.CarListItemEventListener;

import java.util.List;

public class OwnerCarsFragment extends Fragment implements OwnerCarListContract.View {

    private CarListAdapter mAdapter;
    private RecyclerView mCarsListView;

    private OwnerCarsPresenter mPresenter;
    private CarListItemEventListener mEventListener;
    private Toast mCurrentToast;

    public static Fragment newInstance() {
        return new OwnerCarsFragment();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        mEventListener = (CarListItemEventListener) context;
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        mEventListener = (CarListItemEventListener) activity;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mAdapter = new CarListAdapter(getActivity());
        mPresenter = new OwnerCarsPresenter(this);
        mAdapter.subscribe(mPresenter);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_owner_cars, container, false);
        mCarsListView = rootView.findViewById(R.id.owner_cars_list);
        initCarList(mCarsListView);
        return rootView;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mPresenter.refresh();
    }

    @Override
    public void onStart() {
        super.onStart();
        mPresenter.loadItems();
    }

    @Override
    public void onStop() {
        super.onStop();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mPresenter.destroy();
    }

    private void initCarList(RecyclerView listView) {
        listView.setAdapter(mAdapter);
        listView.setLayoutManager(new LinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL, false));
        listView.setItemAnimator(new DefaultItemAnimator());
    }

    @Override
    public void setItems(List<Car> items) {
        mAdapter.insert(items);
    }

    @Override
    public void updateItem(Car car) {
        mAdapter.update(car);
    }

    @Override
    public void removeItem(Car car) {

    }

    @Override
    public void openItem(Car car) {
        if (mEventListener != null) {
            mEventListener.onCarItemClick(car);
        }
    }

    @Override
    public void openDailyPicker(Car car) {
        if (mEventListener != null) {
            mEventListener.onDailyPickerClick(car);
        }
    }

    @Override
    public void openHourlyPicker(Car car) {
        if (mEventListener != null) {
            mEventListener.onHourlyPickerClick(car);
        }
    }

    @Override
    public void openLocationPicker(Car car) {
        if (mEventListener != null) {
            mEventListener.onLocationPickerClick(car);
        }
    }

    @Override
    public void onUnauthorized() {

    }

    @Override
    public void onConnectionLost() {

    }

    @Override
    public void showProgress(boolean show) {
        if (show) {
            mEventListener.onStartLoading();
        } else {
            mEventListener.onStopLoading();
        }
    }

    @Override
    public void showMessage(String message) {
        if (mCurrentToast != null) {
            mCurrentToast.cancel();
        }
        mCurrentToast = Toast.makeText(getActivity(), message, Toast.LENGTH_SHORT);
        mCurrentToast.show();
    }

    @Override
    public void showMessage(@StringRes int messageId) {
        showMessage(getString(messageId));
    }

}

File: app/src/main/java/com/cardee/owner_home/view/OwnerCarsFragment.kt
package com.cardee.owner_home.view

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.support.annotation.StringRes
import android.support.v4.app.Fragment
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.cardee.R
import com.cardee.data_source.Error
import com.cardee.domain.owner.entity.Car
import com.cardee.owner_home.OwnerCarListContract
import com.cardee.owner_home.presenter.OwnerCarsPresenter
import com.cardee.owner_home.view.adapter.CarListAdapter
import com.cardee.owner_home.view.listener.CarListItemEventListener
import kotlinx.android.synthetic.main.fragment_owner_cars.*

class OwnerCarsFragment : Fragment(), OwnerCarListContract.View {

    private var mAdapter: CarListAdapter? = null
    private var mCarsListView: RecyclerView? = null

    private var mPresenter: OwnerCarsPresenter? = null
    private var mEventListener: CarListItemEventListener? = null
    private var mCurrentToast: Toast? = null

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        mEventListener = context as CarListItemEventListener?
    }

    override fun onAttach(activity: Activity?) {
        super.onAttach(activity)
        mEventListener = activity as CarListItemEventListener?
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mAdapter = CarListAdapter(activity as Context?)
        mPresenter = OwnerCarsPresenter(this)
        mAdapter?.subscribe(mPresenter)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater!!.inflate(R.layout.fragment_owner_cars, container, false)
        mCarsListView = rootView.findViewById(R.id.owner_cars_list)
        initCarList(mCarsListView)
        return rootView
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        swipeRefresh.setOnRefreshListener {
            mPresenter?.loadItems()
        }
        mPresenter?.refresh()
    }

    override fun onStart() {
        super.onStart()
        mPresenter?.loadItems()
    }

    override fun onStop() {
        super.onStop()
    }

    override fun onDestroy() {
        super.onDestroy()
        mPresenter?.destroy()
        swipeRefresh?.isEnabled = false
    }

    private fun initCarList(listView: RecyclerView?) {
        listView?.adapter = mAdapter
        listView?.layoutManager = LinearLayoutManager(activity, LinearLayoutManager.VERTICAL, false)
        listView?.itemAnimator = DefaultItemAnimator()
    }

    override fun setItems(items: List<Car>) {
        mAdapter?.insert(items)
    }

    override fun updateItem(car: Car) {
        mAdapter?.update(car)
    }

    override fun removeItem(car: Car) {

    }

    override fun openItem(car: Car) {
        mEventListener?.onCarItemClick(car)
    }

    override fun openDailyPicker(car: Car) {
        mEventListener?.onDailyPickerClick(car)
    }

    override fun openHourlyPicker(car: Car) {
        mEventListener?.onHourlyPickerClick(car)
    }

    override fun openLocationPicker(car: Car) {
        mEventListener?.onLocationPickerClick(car)
    }

    override fun onUnauthorized() {

    }

    override fun onConnectionLost() {
        showMessage(Error.Message.CONNECTION_LOST)
    }

    override fun showProgress(show: Boolean) {
        if (show) {
            mEventListener?.onStartLoading()
        } else {
            mEventListener?.onStopLoading()
            swipeRefresh?.isRefreshing = false
        }
    }

    override fun showMessage(message: String) {
        mCurrentToast?.cancel()
        mCurrentToast = Toast.makeText(activity, message, Toast.LENGTH_SHORT)
        mCurrentToast?.show()
    }

    override fun showMessage(@StringRes messageId: Int) {
        showMessage(getString(messageId))
    }

    companion object {

        fun newInstance(): Fragment {
            return OwnerCarsFragment()
        }
    }

}
